### PR TITLE
GH-35850: [C++] Don't disable optimization with RelWithDebInfo

### LINKED ISSUE #35850: [C++] RelWithDebInfo builds disable optimizations
### Describe the bug, including details regarding any error messages, version, and platform.

RelWithDebInfo builds add `-O0` to compiler flags, effectively disabling optimizations.
This seems to be caused by https://github.com/apache/arrow/pull/15022

### Component(s)

C++

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

We should not disable optimization with RelWithDebInfo. We just add debug information with RelWithDebInfo.

### What changes are included in this PR?

This change prioritizes optimization flags over debug flags.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

Yes.
* Closes: #35850

### CODE REVIEW

File: cpp/cmake_modules/SetupCxxFlags.cmake [hunk 1]
```diff
@@ -642,8 +642,8 @@ if(NOT MSVC)
   string(APPEND CMAKE_CXX_FLAGS_RELEASE "${CXX_RELEASE_FLAGS}")
   string(APPEND CMAKE_C_FLAGS_DEBUG "${DEBUG_FLAGS}")
   string(APPEND CMAKE_CXX_FLAGS_DEBUG "${DEBUG_FLAGS}")
-  string(APPEND CMAKE_C_FLAGS_RELWITHDEBINFO "${C_RELEASE_FLAGS} ${DEBUG_FLAGS}")
-  string(APPEND CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CXX_RELEASE_FLAGS} ${DEBUG_FLAGS}")
+  string(APPEND CMAKE_C_FLAGS_RELWITHDEBINFO "${DEBUG_FLAGS} ${C_RELEASE_FLAGS}")
+  string(APPEND CMAKE_CXX_FLAGS_RELWITHDEBINFO "${DEBUG_FLAGS} ${CXX_RELEASE_FLAGS}")
```
Comment (pitrou) [inline cpp/cmake_modules/SetupCxxFlags.cmake:646]: This works for me, but can you add comment so that we don't inadvertently reintroduce the issue?

**Review by pitrou — CHANGES_REQUESTED** [review]

File: cpp/cmake_modules/SetupCxxFlags.cmake (hunk 1 above)
Comment (kou) [reply cpp/cmake_modules/SetupCxxFlags.cmake:648]: Done.

**Comment by pitrou** [comment]
@github-actions crossbow -g r

**Review by pitrou — APPROVED** [review]
Thanks for the quick fix!

**Comment by pitrou** [comment]
@github-actions crossbow submit r

**Review by mapleFU — APPROVED** [review]

**Comment by pitrou** [comment]
@raulcd I added this to 12.0.1. It's not a must-have but it's a recent regression and it may trip up users.

**Pull request merged**